### PR TITLE
refactoring branch such that ML model related 

### PR DESCRIPTION
**Pull request by vedhasua**
stuff is out of the openpmd-streaming-continual-learning.py and put into a separate directory

### CODE REVIEW

File: scripts/job_hemera.sh [hunk 1]
```diff
@@ -32,4 +32,5 @@ nvidia-smi
 
 BATCH_SIZE="${BATCH_SIZE:-4}"
 mpiexec bash $INSITUML/scripts/ompi_CUDA_VISIBLE_DEVICES_wrapper.sh python $INSITUML/tools/openpmd-streaming-continual-learning.py --io_config $INSITUML/share/configs/io_config_hemera.py  --model_config $INSITUML/share/configs/model_config.py
+# mpiexec bash $INSITUML/scripts/ompi_CUDA_VISIBLE_DEVICES_wrapper.sh python $INSITUML/tools/openpmd-streaming-continual-learning.py --io_config $INSITUML/share/configs/io_config_hemera.py  --model_config $INSITUML/share/configs/model_config.py
```
Comment (jkelling) [inline scripts/job_hemera.sh:35]: This appears to be just a copy of the above.

```suggestion
```

File: share/env/ddp_tested_hemera_env.sh [hunk 2]
```diff
@@ -5,5 +5,5 @@ module load gcc/12.2.0 cuda/12.1 openmpi/4.1.5-cuda121-gdr ucx/1.14.0-gdr \
 	# openpmd/0.15.2-cuda121-blosc2-py3122
 # for (re-)instaling openpmd-api
 export openPMD_USE_MPI=ON
-source /home/kelling/checkout/insitumlNp2Torch26Env/bin/activate
+source /home/pandit52/venvs/Ism/bin/activate
```
Comment (jkelling) [inline share/env/ddp_tested_hemera_env.sh]: No edit wars: do not commit local config changes (at least not into PR) .

File: tools/models/__init__.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+"""
```
Comment (jkelling) [inline tools/models/__init__.py]: Modules need to go into the inSituML code, i.e. into `src/inSituML`.

File: tools/models/architectures.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
+import torch.nn as nn
+from inSituML.loss_functions import EarthMoversLoss
+from inSituML.ks_models import INNModel
```
Comment (jkelling) [inline tools/models/architectures.py]: Note: when moving this int `src/inSituML`, there is no need say `inSituML.`.

File: tools/models/architectures.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+
+class ModelFinal(nn.Module):
```
Comment (jkelling) [inline tools/models/architectures.py]: Having a more generic concept of the model to be trained is great. Note though, this class remains a somewhat specific instance, i.e.  it has encoder, decoder, inner model. Please rename to something more descriptive, to respect this.

File: tools/models/model_factory.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+def get_VAE_encoder_kwargs(io_config, model_config):
+    """Create encoder kwargs dictionary from configs"""
+    return {
+        "ae_config": "non_deterministic",
+        "z_dim": model_config.latent_space_dims,
+        "input_dim": io_config.ps_dims,
+        "conv_layer_config": [16, 32, 64, 128, 256, 608],
+        "conv_add_bn": False,
+        "fc_layer_config": [544],
+    }
```
Comment (jkelling) [inline tools/models/model_factory.py]: This whole dictionary should be defined in model_config.

File: tools/models/model_factory.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+def get_VAE_decoder_kwargs(io_config, model_config):
+    """Create decoder kwargs dictionary from configs"""
+    return {
+        "z_dim": model_config.latent_space_dims,
+        "input_dim": io_config.ps_dims,
+        "initial_conv3d_size": [16, 4, 4, 4],
+        "add_batch_normalisation": False,
+        "fc_layer_config": [1024],
+    }
```
Comment (jkelling) [inline tools/models/model_factory.py]: This whole thing should come from model_config, too.

File: tools/models/model_factory.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+from inSituML.ks_models import INNModel
+
+from inSituML.args_transform import MAPPING_TO_LOSS
+from inSituML.encoder_decoder import Encoder
+from inSituML.encoder_decoder import Conv3DDecoder
+from inSituML.loss_functions import EarthMoversLoss
+from inSituML.networks import VAE
+from models.architectures import ModelFinal
+
+def get_world_size():
```
Comment (jkelling) [inline tools/models/model_factory.py]: This is unrelated to model definitions. The main program should still handle obtaining this value. The function can be moved into a utility module, but it should be a module with parallel-training utilities.

File: tools/models/model_factory.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
+import os, sys
+import torch
+import torch.optim as optim
+from inSituML.utilities import MMD_multiscale, fit, load_checkpoint
+from inSituML.ks_models import INNModel
+
+from inSituML.args_transform import MAPPING_TO_LOSS
+from inSituML.encoder_decoder import Encoder
+from inSituML.encoder_decoder import Conv3DDecoder
```
Comment (jkelling) [inline tools/models/model_factory.py]: Encoder and Decoder class should be declared in model_config (not wholly defined, just imported from model library in general)

File: tools/models/model_factory.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+            loss_latent=MMD_multiscale,
+            loss_backward=MMD_multiscale,
+            lambd_predict=config["lambd_predict"],
+            lambd_latent=config["lambd_latent"],
+            lambd_rev=config["lambd_rev"],
+            zeros_noise_scale=config["zeros_noise_scale"],
+            y_noise_scale=config["y_noise_scale"],
+            hidden_size=config["hidden_size"],
+            activation=config["activation"],
+            num_coupling_layers=config["num_coupling_layers"],
```
Comment (jkelling) [inline tools/models/model_factory.py]: These params should be from a dictionary in model_config. Most of them already are, maybe this could be renamed (more descriptive than "config", maybe "inner_model_config"). The details can then also be left out here, by just passing `**model_config.inner_model_config` to the inner model ctor.

File: tools/models/model_factory.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+            device=rank,
+        )
+
+    # Initialize final model
+    model = ModelFinal(
+        VAE_obj,
+        inner_model,
+        EarthMoversLoss(),
```
Comment (jkelling) [inline tools/models/model_factory.py]: Loss should also be configured in model config later ... but this requires more changes to the model structure, so OK for now.

File: tools/models/model_factory.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    if ("lr_annealingRate" not in config) or config[
+        "lr_annealingRate"
+    ] is None:
+        scheduler = None
+    else:
+        scheduler = torch.optim.lr_scheduler.StepLR(
+            optimizer, step_size=500, gamma=config["lr_annealingRate"]
+        )
+
+    return optimizer, scheduler, model
```
Comment (jkelling) [inline tools/models/model_factory.py]: This is going beyond what should be in a `model_factory` module, because they are parameters to the training, though it is OK for the `load_objects` functions. Rename the file.

Optimizer and LR scheduler should also be configurable in module_config, but with defaults available.

File: tools/openpmd-streaming-continual-learning.py [hunk 13]
```diff
@@ -572,5 +307,6 @@ def run_demo(demo_fn, world_size):
         run_copies(runner=args.runner)
 
 
+
```
Comment (jkelling) [inline tools/openpmd-streaming-continual-learning.py:311]:
```suggestion
```

**Review by jkelling — CHANGES_REQUESTED** [review]
